test/integration/ctl/ctl_create_cr_test.go

```diff
@@ -0,0 +1,211 @@
+			expKeyFilename: cr2Name + ".key",
+		},
+		"conflicting namespaces defined in flag and file": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr3Name},
+			inputNamespace: ns2,
+			keyFilename:    "",
+			expErr:         true,
+			expNamespace:   "",
+			expName:        "",
+			expKeyFilename: "",
+		},
+		"file passed in defines resource other than certificate": {
+			inputFile:      testdataPath + "create_cr_issuer.yaml",
+			inputArgs:      []string{cr4Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expErr:         true,
+			expNamespace:   "",
+			expName:        "",
+			expKeyFilename: "",
+		},
+		"path to file to store private key provided": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr1Name},
+			inputNamespace: ns1,
+			keyFilename:    "test.key",
+			expErr:         false,
+			expNamespace:   ns1,
+			expName:        cr1Name,
+			expKeyFilename: "test.key",
+		},
+	}
+
+	for name, test := range tests {
+		// Run ctl create cr command with input options
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
+				CMClient:         cmCl,
+				RESTConfig:       config,
+				IOStreams:        streams,
+				CmdNamespace:     test.inputNamespace,
+				EnforceNamespace: test.inputNamespace != "",
+				KeyFilename:      test.keyFilename,
+			}
+
+			opts.InputFilenames = []string{test.inputFile}
+
+			err := opts.Run(test.inputArgs)
+
+			if err != nil {
+				if !test.expErr {
+					t.Errorf("got unexpected error when trying to create CR: %v", err)
+				}
+				t.Logf("got an error, which was expected, details: %v", err)
+				return
+			} else {
+				// got no error
+				if test.expErr {
+					t.Errorf("expected but got no error when creating CR")
+				}
+			}
+
+			// Finished creating CR, check if everything is expected
+			crName := test.inputArgs[0]
+			gotCr, err := cmCl.CertmanagerV1alpha2().CertificateRequests(test.inputNamespace).Get(ctx, crName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if gotCr.Namespace != test.expNamespace {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
+				t.Errorf("CR created in unexpected Namespace, expected: %s, actual: %s", test.expNamespace, gotCr.Namespace)
+			}
+
+			if gotCr.Name != test.expName {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
+				t.Errorf("CR created has unexpected Name, expectedL %s, actualL %s", test.expName, gotCr.Name)
+			}
+
+			// Check the file where the private key is stored
+			keyData, err := ioutil.ReadFile(test.expKeyFilename)
+			if err != nil {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
```

Comment:
We don’t need to explicitly/manually delete things everywhere, this will be thrown away at the end of the test case anyway and makes the test harder to read/follow 🙂